Feature/json schema article (#6787)
pubkey/rxdb: docs-src/docs/article-json-schema.md

# How RxDB benefits from JSON Schema to build its NoSQL Database

[RxDB](https://rxdb.info/) is a NoSQL database designed primarily for client-side JavaScript applications, focusing on the [offline-first approach](https://rxdb.info/offline-first.html), multi-tab operations, and data replication. It seamlessly integrates with modern web and mobile app frameworks, supports various storage backends, and provides a plugin-based architecture to handle encryption, compression, type generation, and more.

One of RxDB's standout architectural choices is its reliance on JSON Schema to define data models for documents in collections. Instead of inventing a custom schema language, RxDB adopts a standard that is already well-established in the broader JavaScript community. By doing so, RxDB solves several major hurdles that often arise when developers are introduced to new databases:

> - **Familiarity**: Most JavaScript developers have already encountered JSON-based schema definitions, either via OpenAPI or other tooling. Thus, they can quickly understand and adopt RxDB's schemas without investing time in learning a proprietary format.

Reviewer: ```suggestion
- **Familiarity**: Most JavaScript developers have already encountered JSON Schema, either via OpenAPI or other tooling. Thus, they can quickly understand and adopt RxDB's schemas without investing time in learning a new schema language.
```

- **Tooling ecosystem**: JSON Schema boasts a robust set of libraries, validators, and code generators. RxDB leverages these rather than maintaining its own schema validation library, giving teams the freedom to pick what works best for them.

- **Long-term maintainability**: Because JSON Schema is recognized as a standard, future updates and ecosystem support are more reliable than a custom, one-off solution.

By building on JSON Schema, RxDB has a foundation that makes schema design, data validation, and typing straightforward, making it easier for developers to build robust, safe applications in production.

<center>
    <a href="https://rxdb.info/">
        <img src="https://rxdb.info/files/logo/rxdb_javascript_database.svg" alt="JavaScript Database" width="220" />
    </a>
</center>

## How RxDB uses JSON Schema

> While RxDB reuses the core JSON Schema specification, it also extends that schema to introduce RxDB-specific features. Like in other NoSQL databases, you manually define which fields to encrypt, which ones to index, and how to interpret specific fields for queries. RxDB puts these configurations directly into the JSON Schema with additional properties:

Reviewer: ```suggestion
While RxDB adopts the JSON Schema Core and Validation specifications, it also extends it to introduce RxDB-specific functionality. Like in other NoSQL databases, you can manually define which fields to encrypt, which ones to index, and how to interpret specific fields for queries. RxDB enables these configurations as custom JSON Schema keywords:
```

> - `primaryKey`: Specifies which field in the document serves as the primary key.

Reviewer: Interesting stuff btw. One line of work we did in the past was helping Oracle define a standard "database" extension to JSON Schema: https://github.com/json-schema-org/vocab-database/blob/main/database.md (which they implement at Oracle). I wonder if that same database vocabulary could be a good fit for RxDB and if it could be extended for the things you do that they don't!

Reviewer: Yeah, maybe that's simpler, just to avoid dealing with vocabularies, though if you ever want to do it properly, happy to help you out in all of it. Just let me know!

Author: Will come back to you. But this would be a breaking change. So earliest would be end of this year because we had a major release some weeks ago already.

- `indexes`: Defines which fields (or combination of fields) RxDB indexes. You can have single-field indexes or compound indexes.

> - `version`: Indicates the version of the schema. Whenever you change your schema, you must increment this version so RxDB can handle migrations or other adjustments. This is important because data migration on a client-side database can be tricky when you have many clients out there that update your app at different points in time.

Reviewer: I think it's OK for the blog post. For your later consideration, the current "standard" way of doing versioning is with `$id`, and making the version part of that. For cases like RxDB, URNs or Tag URIs can be interesting!

Author: I will think about these. RxDB is so complex already and having the user to learn about the URI instead of a version `number` might be more confusing than what is solves.

Reviewer: Maybe the version field can go in a wrapper around the schema too? Like you mentioned here: https://github.com/pubkey/rxdb/pull/6787/files#r1934107741

Author: Yes that would have to me in this config then also.

> - `encryption`: Specifies which fields should be stored in an encrypted form. This is useful for sensitive data that you do not want to store in plaintext on the client's device.

Reviewer: Can we mention `keyCompression` here?

> RxDB also supports a [compression plugin](https://rxdb.info/key-compression.html) that uses a "compression table." This is essentially a lookup table derived from your JSON Schema which assigns shorter keys or transforms fields so that the stored data becomes more compact. By analyzing the schema, the plugin understands which fields appear repeatedly and can replace them with shorter tokens. Remarkably, RxDB can still query the data in its compressed form. This leads to performance improvements, especially in environments where local storage space is limited.

Reviewer: Very nice! I would love to learn more about this

> Below is a sample RxDB schema that demonstrates how standard JSON Schema fields combine with RxDB's custom properties:

Reviewer: ```suggestion
Below is a sample RxDB schema that demonstrates how standard JSON Schema vocabularies combine with RxDB's custom extensions:
```

```ts

> const mySchema = {

Reviewer: Another point of contention you might get is that if schemas doesn't declare `$schema`, the version of JSON Schema you are using is ambiguous and can lead to unexpected problems. If you don't make use of `$schema` right now, can you update the post to very clearly state which dialect of JSON Schema (i.e. JSON Schema 2020-12) RxDB assumes?

Author: RxDB does not assume a specific draft. What it accepts and what it does not accept, depends on which validator the user picks.

    primaryKey: 'id',
    type: 'object',
    properties: {
        id: {
            type: 'string',
            maxLength: 100
        },
        firstName: {
            type: 'string'
        },
        lastName: {
            type: 'string'
        },
        secret: {
            type: 'string'
        }
    },
    additionalProperties: false,
    // RxDB specific fields:
    primaryKey: 'id',
    version: 0,
    keyCompression: true,
    required: ['firstName', 'lastName', 'id'],
    indexes: [
        'firstName', // single-field index
        ['firstName', 'lastName'] // compound index
    ],
    encrypted: ['secret']
}
```

## Restrictions

> Although RxDB aims to remain compatible with JSON Schema, it does impose some extra restrictions:

Reviewer: ```suggestion
Although RxDB aims to remain fully spec compliant with JSON Schema, it does impose some extra restrictions:
```

The schema requires top-level field names to match the regex `^[a-zA-Z][a-zA-Z0-9_]*[a-zA-Z0-9]$`. This rule ensures that property names are valid JavaScript identifiers.

At the top level of the schema, `additionalProperties` must be set to `false`. This prevents silently introducing new fields that could potentially clash with built-in RxDocument methods (like `.toJSON()`). By making you list out each property, RxDB ensures you can't unknowingly overwrite or conflict with standard methods.

> RxDB does not allow `$ref` to other files or external schema fragments. The goal is to keep schema loading fast and self-contained. If a schema needs to be composed of multiple parts, you must combine them at build time or otherwise ensure they are merged before passing them to RxDB at runtime. This approach also prevents any network calls or asynchronous fetches that could slow down your application's startup.

Reviewer: Is RxDB on JSON Schema 2020-12? If so, also mention `$dynamicRef` alongside `$ref`?

Author: I will add `$dynamicRef`

## Inferring Document Types with TypeScript

JSON Schema is not just for validation and structural guarantees, it can also help generate or infer TypeScript types. In many projects, developers rely on tools like [json-schema-to-typescript](https://github.com/bcherny/json-schema-to-typescript) to produce `.d.ts` files or interface definitions from schema files at build time. However, that requires a separate build step, which slows down your workflow.

> To improve developer experience, RxDB offers a [built-in way](https://rxdb.info/tutorials/typescript.html) to infer the document's TypeScript type from the schema during runtime. This is helpful because you get Immediate feedback in your IDE: As soon as you update the schema, TypeScript picks up the changes. You'll see type errors in your code if you attempt to use fields that are no longer valid or if you forget to include newly required fields.

Reviewer: ```suggestion
To improve developer experience, RxDB offers a [built-in way](https://rxdb.info/tutorials/typescript.html) to infer the document's TypeScript type from the schema during runtime. This is helpful because you get immediate feedback in your IDE: As soon as you update the schema, TypeScript picks up the changes. You'll see type errors in your code if you attempt to use fields that are no longer valid or if you forget to include newly required fields.
```

const schemaTyped = toTypedRxJsonSchema(heroSchemaLiteral);

// Extract the document type from the typed schema
export type HeroDocType = ExtractDocumentTypeFromTypedRxJsonSchema<typeof schemaTyped>;

// Create a typed RxJsonSchema
export const heroSchema: RxJsonSchema<HeroDocType> = heroSchemaLiteral;

// Example usage: create a typed collection
const myCollection: RxCollection<HeroDocType> = db.heroes;
```

## Different JSON Schema Validators

In order to validate whether a given document complies with the schema, RxDB uses pluggable validators. Historically, RxDB supports/supported:

- [ajv](https://ajv.js.org/)
- [z-schema](https://github.com/zaggino/z-schema)
- [is-my-json-valid](https://github.com/mafintosh/is-my-json-valid) (now deprecated in RxDB because it incorrectly handles large and decimal numbers)

No single validator is perfect; each has its trade-offs in performance, build size, and correctness edge cases. Some libraries rely on `eval()` or `new Function`, which can break in strict Content-Security-Policy (CSP) environments. Others have issues with large integers. RxDB encourages you to pick a validator that matches your app's needs and environment constraints.

> One important consideration is that each validator has a unique format for its **error messages**. If your application inspects validation errors and makes decisions (e.g., showing descriptive warnings to the user), be aware that switching validators later can require extensive code changes.

Reviewer: The JSON Schema Core specification describes "Standard Output Formats" to fix exactly this issue: unifying how different validators report errors, though not all validators implement them (AJV doesn't). Might be worth commenting that you are aware of that feature and that if all validators did follow the spec by heart, this wouldn't be a problem?

See: https://json-schema.org/draft/2020-12/json-schema-core#name-output-formatting

> ## Performance comparison of validators

Reviewer: I noticed some titles are capitalised (like the one before) but some are not (like this one). Can you revise them for consistency?

Performance is a critical factor in deciding whether to validate documents at runtime, especially in production environments. The following tables illustrate a basic comparison of initialization time (time-to-first-insert) and bulk insertion speed for different validators on two RxDB storages

> The RxDB team ran performance benchmarks using two storage options:

Author: I will add this.

## Should JSON Schema Validation Be Used in Production?

Many teams limit JSON Schema validation to development builds to avoid performance overhead in production. However, if your application deals with highly sensitive or mission-critical data, keeping validation enabled ensures data integrity and can prevent costly errors, despite the added CPU and bundle-size costs. Ultimately, the choice depends on your performance targets and  the risk of invalid data.

### Running the Validation in a WebWorker

If you must keep validation enabled in production but you have to ensure that your UI does not lack during validation, you might consider the [RxDB WebWorker plugin](https://rxdb.info/rx-storage-worker.html). This plugin runs the RxDB storage & validation in a separate Web Worker, offloading the main UI thread. While it won't reduce the absolute time spent on validation, it can help maintain a smooth UI by preventing blocking operations on the main thread.

## Learnings

Over time, RxDB has evolved its usage of JSON Schema, learning from real production experiences and feedback from the community. Here are some key takeaways:

> - Avoid inlined `required` fields: Some validators let you write `"required": true` inside the property definition, but the official JSON Schema specification requires you to declare an array of required fields at the parent object level. Following the standard from the start avoids confusion and ensures broader validator compatibility.

Reviewer: This is historical. JSON Schema Draft 3 used the boolean form of `required`. The issue here really touches on what I mentioned before: most schema authors avoid `$schema`, thus set no versions, and everybody gets confused about what version of JSON Schema they are actually using. Can you revise this paragraph, maybe rephrasing it as some validators being obsolete? (as `required: true` was indeed valid ages ago!)

> - Keep Custom Fields at the Top Level: Originally, RxDB allowed custom definitions (`index`, `encrypted`, etc.) to appear deeply nested. This caused performance hits because the library had to traverse large schema objects to find them. By placing these fields at the top level, RxDB can parse and apply them much faster, improving startup times.

Reviewer: I now wonder if these custom fields even need to be part of the schema. Maybe you have a JSON object with these properties and a `schema` property? Then you don't have to deal with the complexity of creating a custom JSON Schema vocabulary, etc

Author: I will think about this. Maybe RxDB should move them to somewhere else.

> - Error messages are not standardized: Each validator produces a different structure for error messages. If your app logic inspects these errors, you risk partial or complete rewrites if you ever switch validators. Decide early on which validator meets your needs and plan on sticking with it long-term.

Reviewer: This is a huge one we are constantly battling against. Can you mention the Standard Output Formats I touched on before? If you can express your desire for validators to comply, that would be wonderful!

Author: Yes, I add this.

const schemaTyped = toTypedRxJsonSchema(heroSchemaLiteral);

// Extract the document type from the typed schema
export type HeroDocType = ExtractDocumentTypeFromTypedRxJsonSchema<typeof schemaTyped>;

// Create a typed RxJsonSchema
export const heroSchema: RxJsonSchema<HeroDocType> = heroSchemaLiteral;

// Example usage: create a typed collection
const myCollection: RxCollection<HeroDocType> = db.heroes;
```

## Different JSON Schema Validators

In order to validate whether a given document complies with the schema, RxDB uses pluggable validators. Historically, RxDB supports/supported:

- [ajv](https://ajv.js.org/)
- [z-schema](https://github.com/zaggino/z-schema)
- [is-my-json-valid](https://github.com/mafintosh/is-my-json-valid) (now deprecated in RxDB because it incorrectly handles large and decimal numbers)

No single validator is perfect; each has its trade-offs in performance, build size, and correctness edge cases. Some libraries rely on `eval()` or `new Function`, which can break in strict Content-Security-Policy (CSP) environments. Others have issues with large integers. RxDB encourages you to pick a validator that matches your app's needs and environment constraints.

> Notice that RxDB itself does not define a specific JSON Schema draft. Instead it fully depends on the validation plugin is use, which parts of the JSON schema spec are accepted.

Reviewer: This paragraph doesn't read very well. Maybe:

> Note that RxDB itself does not enforce the use of a specific JSON Schema dialect. Instead, the range of supported dialects is determined by the JSON Schema validation plugin in use.

Also BTW, we tend to say "dialect" instead of "draft". The "draft" terminology is just old IETF annoyance

Author: I replaced "draft" with dialect everywhere.

> Notice that RxDB itself does not define a specific JSON Schema draft. Instead it fully depends on the validation plugin is use, which parts of the JSON schema spec are accepted.

Reviewer: This paragraph doesn't read very well. Maybe:

> Note that RxDB itself does not enforce the use of a specific JSON Schema dialect. Instead, the range of supported dialects is determined by the JSON Schema validation plugin in use.

Also BTW, we tend to say "dialect" instead of "draft". The "draft" terminology is just old IETF annoyance

> Notice that RxDB itself does not define a specific JSON Schema draft. Instead it fully depends on the validation plugin is use, which parts of the JSON schema spec are accepted.

Reviewer: This paragraph doesn't read very well. Maybe:

> Note that RxDB itself does not enforce the use of a specific JSON Schema dialect. Instead, the range of supported dialects is determined by the JSON Schema validation plugin in use.

Also BTW, we tend to say "dialect" instead of "draft". The "draft" terminology is just old IETF annoyance

> Notice that RxDB itself does not define a specific JSON Schema draft. Instead it fully depends on the validation plugin is use, which parts of the JSON schema spec are accepted.

Reviewer: This paragraph doesn't read very well. Maybe:

> Note that RxDB itself does not enforce the use of a specific JSON Schema dialect. Instead, the range of supported dialects is determined by the JSON Schema validation plugin in use.

Also BTW, we tend to say "dialect" instead of "draft". The "draft" terminology is just old IETF annoyance

> One important consideration is that each validator has a unique format for its **error messages**. If your application inspects validation errors and makes decisions (e.g., showing descriptive warnings to the user), be aware that switching validators later can require extensive code changes. In theory this is solved by the JSON schema specification with the [output formatting](https://json-schema.org/draft/2020-12/json-schema-core#name-output-formatting) but for now this is not implemented in the listed schema validators.

Reviewer: ```suggestion
One important consideration is that each validator has a unique format for its **error messages**. If your application inspects validation errors and makes decisions (e.g., showing descriptive warnings to the user), be aware that switching validators later can require extensive code changes. In theory this is solved by the JSON Schema specification by the use of the [Standard Output](https://json-schema.org/draft/2020-12/json-schema-core#name-output-formatting) format, but this is not implemented in the listed schema validators.
```

## Performance Comparison of Validators

Performance is a critical factor in deciding whether to validate documents at runtime, especially in production environments. The following tables illustrate a basic comparison of initialization time (time-to-first-insert) and bulk insertion speed for different validators on two RxDB storages

> The RxDB team ran performance benchmarks using two storage options on an ubuntu 24 machine with Chrome version `131.0.6778.85`. The testing machine has 32 core `13th Gen Intel(R) Core(TM) i9-13900HX` CPU so notice that your average users' device will be way slower.

Reviewer: ```suggestion
The RxDB team ran performance benchmarks using two storage options on an Ubuntu 24.04 machine with Chrome version `131.0.6778.85`. The testing machine has 32 core `13th Gen Intel(R) Core(TM) i9-13900HX` CPU.
```

## Should JSON Schema Validation Be Used in Production?

Many teams limit JSON Schema validation to development builds to avoid performance overhead in production. However, if your application deals with highly sensitive or mission-critical data, keeping validation enabled ensures data integrity and can prevent costly errors, despite the added CPU and bundle-size costs. Ultimately, the choice depends on your performance targets and  the risk of invalid data.

### Running the Validation in a WebWorker

If you must keep validation enabled in production but you have to ensure that your UI does not lack during validation, you might consider the [RxDB WebWorker plugin](https://rxdb.info/rx-storage-worker.html). This plugin runs the RxDB storage & validation in a separate Web Worker, offloading the main UI thread. While it won't reduce the absolute time spent on validation, it can help maintain a smooth UI by preventing blocking operations on the main thread.

## Learnings

Over time, RxDB has evolved its usage of JSON Schema, learning from real production experiences and feedback from the community. Here are some key takeaways:

> - Avoid inlined `required` fields: Some validators let you write `"required": true` inside the property definition because that was part of JSON Schema Draft 3. But the newer JSON Schema drafts require you to declare an array of required fields at the parent object level. Following the standard from the start avoids confusion and ensures broader validator compatibility.

Reviewer: I think it reads better now, but we probably should say to "avoid inlined required fields". Previously in the article, you say the JSON Schema dialect in use depends on the user plugin, so in theory some of them might use a plugin that is intentionally Draft 3, and thus this advice wouldn't make sense.

Maybe you can rephrase the advise as: make sure you know which JSON Schema dialect is in use to avoid confusion, citing inlined `required` and Draft 3 as an example of such confusion?
